yarnpkg/yarn
# Add optional offline mirror pruning

**Summary**

This change is an implementation of yarnpkg/rfcs#49, which stems from #2109.





**Test plan**

I tested the feature locally with two scenarios where `.yarnrc` has `yarn-offline-mirror-pruning` set to `true`:

* Upgrading a package:
    * `$ yarn add is-positive-integer@1.0.0`
    * Check that the offline mirror has the `is-positive-integer` tarball, along with all its dependencies
    * `$ yarn upgrade is-positive-integer@1.1.1`
    * Check that the offline mirror only has the new version of `is-positive-integer`, which doesn't have any dependencies
* Removing a package:
    * `$ yarn add left-pad right-pad`
    * Check that the offline mirror has both the `left-pad` and `right-pad` tarballs
    * `$ yarn remove left-pad`
    * Check that the offline mirror only has `right-pad`

I also added two Jest test cases, one for upgrade and one for remove. They verify that sub-dependencies are also pruned.


For documentation, I'd be happy to submit a pull request for that as well, but I don't see any for the offline mirror feature on the main Yarn website. I've just been referencing [this blog post](https://yarnpkg.com/blog/2016/11/24/offline-mirror/).
